In [1]:
from google.colab import drive
import sys
import pandas as pd
import numpy as np

# **Load Data**

In [2]:
# Mount Google Drive
drive.mount('/content/drive')

# Get the absolute path of the current folder
abspath_curr = '/content/drive/My Drive/'

Mounted at /content/drive


In [3]:
transactions=pd.read_csv(abspath_curr+'transactions.csv',index_col=[0])
transactions

accountNumber  customerId  ...  expirationDateKeyInMatch  isFraud
0           737265056   737265056  ...                     False    False
1           737265056   737265056  ...                     False    False
2           737265056   737265056  ...                     False    False
3           737265056   737265056  ...                     False    False
4           830329091   830329091  ...                     False    False
...               ...         ...  ...                       ...      ...
786358      732852505   732852505  ...                     False    False
786359      732852505   732852505  ...                     False    False
786360      732852505   732852505  ...                     False    False
786361      732852505   732852505  ...                     False    False
786362      732852505   732852505  ...                     False    False

[786363 rows x 29 columns]

# **Data Wrangling**

##Set the transaction key

In [4]:
# Set the transaction key
def concat(x):
    y=str('T') + str(x)
    return y

transactions.reset_index(drop=False,inplace=True)
transactions['transactionKey']=transactions['index'].apply(lambda x: concat(x))
transactions.drop(['index'],axis=1,inplace=True)

##**Reversed Transactions**


Assumptions

1.  I assume the reversed transactions contain both purchase and reversal transaction types.
2.  I assume the reversed transactions have the same transaction amounts beween two transaction types.



In [5]:
# Find the transactions with "Purchase" and "Reversal" 
transa=transactions[(transactions['transactionType']== 'PURCHASE')|(transactions['transactionType']=='REVERSAL')][['customerId','merchantName','transactionAmount','transactionType','transactionKey']]
transa

customerId         merchantName  ...  transactionType transactionKey
0        737265056                 Uber  ...         PURCHASE             T0
1        737265056          AMC #191138  ...         PURCHASE             T1
2        737265056           Play Store  ...         PURCHASE             T2
3        737265056           Play Store  ...         PURCHASE             T3
4        830329091  Tim Hortons #947751  ...         PURCHASE             T4
...            ...                  ...  ...              ...            ...
786358   732852505                 Lyft  ...         PURCHASE        T786358
786359   732852505             hulu.com  ...         PURCHASE        T786359
786360   732852505                 Lyft  ...         PURCHASE        T786360
786361   732852505          walmart.com  ...         PURCHASE        T786361
786362   732852505                 Uber  ...         PURCHASE        T786362

[765496 rows x 5 columns]

In [7]:
#Group data by 'customerId' and 'merchantName' 
#Select transactions in both same customer and merchant
count_ = transactions.groupby(['customerId','merchantName']).count()
count_Amount = count_[count_['transactionAmount']>1].reset_index()[['customerId','merchantName']]
count_Amount

customerId                   merchantName
0       100088067                    AMC #552863
1       100088067                    AMC #724446
2       100088067                     AMC #79863
3       100088067                   Apple iTunes
4       100088067  Blue Mountain Online Services
...           ...                            ...
79848   999283629                      sears.com
79849   999283629                    staples.com
79850   999283629                     target.com
79851   999283629                    walmart.com
79852   999389635                     Play Store

[79853 rows x 2 columns]

In [9]:
# Combine the data 
records = pd.merge(count_Amount, transa, left_on=['customerId','merchantName'],right_on=['customerId','merchantName'])
records 

customerId merchantName  ...  transactionType transactionKey
0        100088067  AMC #552863  ...         PURCHASE        T541951
1        100088067  AMC #552863  ...         PURCHASE        T541958
2        100088067  AMC #724446  ...         PURCHASE        T541909
3        100088067  AMC #724446  ...         PURCHASE        T541946
4        100088067  AMC #724446  ...         PURCHASE        T541954
...            ...          ...  ...              ...            ...
640240   999389635   Play Store  ...         PURCHASE        T108110
640241   999389635   Play Store  ...         PURCHASE        T108112
640242   999389635   Play Store  ...         PURCHASE        T108113
640243   999389635   Play Store  ...         PURCHASE        T108114
640244   999389635   Play Store  ...         PURCHASE        T108115

[640245 rows x 5 columns]

In [11]:
# Group data by the same transaction amounts
# Drop rows with only one transaction
count_records = records.groupby(['customerId','merchantName','transactionAmount']).count()
count_data = count_records[count_records['transactionType']>1].reset_index()[['customerId','merchantName','transactionAmount']]
count_data

customerId          merchantName  transactionAmount
0       100088067            AMC #79863              22.32
1       100088067          Apple iTunes               3.96
2       100088067         Fresh Flowers             411.35
3       100088067     Shell Gas #256420              60.41
4       100108752     Shell Gas #494785              45.35
...           ...                   ...                ...
32438   999283629  Franks Sandwitch Bar             150.58
32439   999283629                  Lyft              96.56
32440   999283629               abc.com              13.04
32441   999283629             apple.com             407.23
32442   999389635            Play Store               4.62

[32443 rows x 3 columns]

In [12]:
# Map the data between count_data and records
card_data = pd.merge(count_data, records, left_on=['customerId','merchantName','transactionAmount'],right_on=['customerId','merchantName','transactionAmount'])
card_data.drop_duplicates(subset=['customerId','merchantName','transactionAmount','transactionType'],keep='first',inplace=True)
card_data

customerId       merchantName  ...  transactionType transactionKey
0        100088067         AMC #79863  ...         PURCHASE        T541963
1        100088067         AMC #79863  ...         REVERSAL        T541964
2        100088067       Apple iTunes  ...         PURCHASE        T541919
10       100088067      Fresh Flowers  ...         PURCHASE        T541955
12       100088067  Shell Gas #256420  ...         PURCHASE        T541953
...            ...                ...  ...              ...            ...
101049   999283629               Lyft  ...         REVERSAL        T670052
101050   999283629            abc.com  ...         PURCHASE        T670063
101056   999283629          apple.com  ...         PURCHASE        T670084
101057   999283629          apple.com  ...         REVERSAL        T670091
101058   999389635         Play Store  ...         PURCHASE        T108108

[50214 rows x 5 columns]

In [13]:
reversal_data = card_data.groupby(['customerId','merchantName','transactionAmount']).count()
reversal_ = reversal_data[reversal_data['transactionType']>1].reset_index()[['customerId','merchantName','transactionAmount','transactionKey']]
reversal_.drop(columns=['transactionKey'],inplace=True)
reversal_


customerId          merchantName  transactionAmount
0       100088067            AMC #79863              22.32
1       100328049                  Lyft              43.74
2       100328049     McDonalds #423357             284.97
3       100737756      34th BBQ #436606              93.67
4       100737756   Best Bistro #262998             501.29
...           ...                   ...                ...
17766   999258704               gap.com              72.93
17767   999283629        Delta Airlines             247.40
17768   999283629  Franks Sandwitch Bar             150.58
17769   999283629                  Lyft              96.56
17770   999283629             apple.com             407.23

[17771 rows x 3 columns]

###Identify reversed transactions





In [14]:
# Get the reversal transactions
reversal_transactions = pd.merge(reversal_, card_data, left_on = ['customerId','merchantName','transactionAmount'],right_on=['customerId','merchantName','transactionAmount'])
reversal_transactions.drop_duplicates(subset=['customerId','merchantName','transactionAmount','transactionType'],keep='first',inplace=True)
reversal_transactions

customerId          merchantName  ...  transactionType transactionKey
0       100088067            AMC #79863  ...         PURCHASE        T541963
1       100088067            AMC #79863  ...         REVERSAL        T541964
2       100328049                  Lyft  ...         PURCHASE         T87919
3       100328049                  Lyft  ...         REVERSAL         T87920
4       100328049     McDonalds #423357  ...         PURCHASE         T87935
...           ...                   ...  ...              ...            ...
35537   999283629  Franks Sandwitch Bar  ...         REVERSAL        T670109
35538   999283629                  Lyft  ...         PURCHASE        T670051
35539   999283629                  Lyft  ...         REVERSAL        T670052
35540   999283629             apple.com  ...         PURCHASE        T670084
35541   999283629             apple.com  ...         REVERSAL        T670091

[35542 rows x 5 columns]



> **total number of transactions for the reversed transactions is 17771**



In [15]:
reversal_transactions['transactionType'].value_counts()

PURCHASE    17771
REVERSAL    17771
Name: transactionType, dtype: int64



>   **total dollar amount for the reversed transactions corresponds to $2667031.44**



In [16]:
reversal_transactions.groupby(['transactionType']).sum()['transactionAmount']

transactionType
PURCHASE    2667031.44
REVERSAL    2667031.44
Name: transactionAmount, dtype: float64

In [17]:
transactions['transactionType'].value_counts()

PURCHASE                745193
REVERSAL                 20303
ADDRESS_VERIFICATION     20169
Name: transactionType, dtype: int64

* There are 2532 reversal transactions which I can not identify.

* I think there are two reasons: 1)The amounts in purchase and reversal can be different.2) Data may be incomplete.





 
 

In [18]:
transactions.groupby(['transactionType']).sum()['transactionAmount']

transactionType
ADDRESS_VERIFICATION    0.000000e+00
PURCHASE                1.047903e+08
REVERSAL                2.821792e+06
Name: transactionAmount, dtype: float64



> ** Total dollar amount for the reversal transactions which I can not identified corresponds to $154450.11.**



In [20]:
total_=transa[(transa['transactionType']=='REVERSAL')][['customerId','merchantName','transactionAmount','transactionType']]
reseveral_su=reversal_transactions[(reversal_transactions['transactionType']=='REVERSAL')]
reserversal_left=reseveral_su.append(total_)
reserversal_left.drop_duplicates(keep=False,inplace=True)
reserversal_left.reset_index()
reserversal_left.groupby(['transactionType']).sum()['transactionAmount']

transactionType
REVERSAL    154450.11
Name: transactionAmount, dtype: float64



###Delete the reversed transactions



In [19]:
credit_card_no_reversed=transactions[~(transactions.transactionKey.isin(reversal_transactions.transactionKey))]
credit_card_no_reversed

accountNumber  customerId  ...  isFraud  transactionKey
0           737265056   737265056  ...    False              T0
1           737265056   737265056  ...    False              T1
2           737265056   737265056  ...    False              T2
3           737265056   737265056  ...    False              T3
4           830329091   830329091  ...    False              T4
...               ...         ...  ...      ...             ...
786358      732852505   732852505  ...    False         T786358
786359      732852505   732852505  ...    False         T786359
786360      732852505   732852505  ...    False         T786360
786361      732852505   732852505  ...    False         T786361
786362      732852505   732852505  ...    False         T786362

[750821 rows x 30 columns]

##**Multi-Wipe Transactions**


Assumption:

 I assume the multi-wipe transactions happened at the same day.




In [20]:
Multi_wipe_data = credit_card_no_reversed[(credit_card_no_reversed['transactionType'] == 'PURCHASE')][['customerId','transactionAmount','merchantName','transactionType','transactionDateTime','transactionKey']]
Multi_wipe_data['customerId'] = Multi_wipe_data['customerId'].astype(dtype='int64')
Multi_wipe_data['transaction_DateTime'] = Multi_wipe_data['transactionDateTime'].apply(lambda x:x.split('T')[0])
Multi_wipe_data.drop(labels='transactionDateTime',axis=1,inplace=True)
Multi_wipe_data

customerId  transactionAmount  ... transactionKey transaction_DateTime
0        737265056              98.55  ...             T0           2016-08-13
1        737265056              74.51  ...             T1           2016-10-11
2        737265056               7.47  ...             T2           2016-11-08
3        737265056               7.47  ...             T3           2016-12-10
4        830329091              71.18  ...             T4           2016-03-24
...            ...                ...  ...            ...                  ...
786358   732852505             119.92  ...        T786358           2016-12-22
786359   732852505              18.89  ...        T786359           2016-12-25
786360   732852505              49.43  ...        T786360           2016-12-27
786361   732852505              49.89  ...        T786361           2016-12-29
786362   732852505              72.18  ...        T786362           2016-12-30

[727422 rows x 6 columns]

In [21]:
# Group data by the same date, same customer and same merchant
mul_count = Multi_wipe_data.groupby(['customerId','merchantName','transaction_DateTime']).count()
mul_count = mul_count.reset_index()

In [22]:
# Select the transactions 
mutiple_counts = mul_count[mul_count['transactionAmount']>1][['customerId','merchantName','transaction_DateTime','transactionKey']]
mutiple_counts.drop(columns=['transactionKey'],inplace=True)
mutiple_counts

customerId               merchantName transaction_DateTime
42       100088067              Fresh Flowers           2016-10-16
229      100737756  1st Sandwitch Bar #396252           2016-05-23
245      100737756           34th BBQ #166379           2016-07-10
308      100737756                Franks Deli           2016-01-18
362      100737756    South Steakhouse #73819           2016-07-02
...            ...                        ...                  ...
658411   999257059        Oklahoma City Times           2016-03-14
658452   999257059           Wall Street News           2016-01-14
658465   999257059          Wall Street Times           2016-02-25
658473   999257059          Wall Street Times           2016-12-17
658652   999283629                alibaba.com           2016-12-07

[36640 rows x 3 columns]

###Identify multi-wipe transactions

In [23]:
Multiwipe=pd.merge(mutiple_counts,Multi_wipe_data,left_on=['customerId','merchantName','transaction_DateTime'], right_on=['customerId','merchantName','transaction_DateTime'])
Multiwipe

customerId               merchantName  ... transactionType  transactionKey
0        100088067              Fresh Flowers  ...        PURCHASE         T541955
1        100088067              Fresh Flowers  ...        PURCHASE         T541956
2        100737756  1st Sandwitch Bar #396252  ...        PURCHASE         T151492
3        100737756  1st Sandwitch Bar #396252  ...        PURCHASE         T151493
4        100737756           34th BBQ #166379  ...        PURCHASE         T151517
...            ...                        ...  ...             ...             ...
105333   999257059          Wall Street Times  ...        PURCHASE         T569679
105334   999257059          Wall Street Times  ...        PURCHASE         T569799
105335   999257059          Wall Street Times  ...        PURCHASE         T569800
105336   999283629                alibaba.com  ...        PURCHASE         T670115
105337   999283629                alibaba.com  ...        PURCHASE         T670116

[105338 rows x 6 columns]

In [24]:
# Get the normal transactions in Multi-wipe transactions
Multiwipe_Normal=Multiwipe.drop_duplicates(subset=['customerId','merchantName','transaction_DateTime'],keep='first')
Multiwipe_Normal

customerId               merchantName  ... transactionType  transactionKey
0        100088067              Fresh Flowers  ...        PURCHASE         T541955
2        100737756  1st Sandwitch Bar #396252  ...        PURCHASE         T151492
4        100737756           34th BBQ #166379  ...        PURCHASE         T151517
6        100737756                Franks Deli  ...        PURCHASE         T151426
9        100737756    South Steakhouse #73819  ...        PURCHASE         T151512
...            ...                        ...  ...             ...             ...
105328   999257059        Oklahoma City Times  ...        PURCHASE         T569684
105330   999257059           Wall Street News  ...        PURCHASE         T569653
105332   999257059          Wall Street Times  ...        PURCHASE         T569678
105334   999257059          Wall Street Times  ...        PURCHASE         T569799
105336   999283629                alibaba.com  ...        PURCHASE         T670115

[36640 rows x 6 columns]

**Total number of transactions for the multi-swipe transactions is 68698.**

In [88]:
105338-36640

68698

In [25]:
Multiwipe_Du=pd.merge(Multiwipe_Normal,Multiwipe,on=['customerId','merchantName','transaction_DateTime','transactionKey','transactionAmount','transactionType'],how='right',indicator=True)
Multiwipe_Du=Multiwipe_Du[(Multiwipe_Du['_merge']== 'right_only')][['customerId','merchantName','transaction_DateTime','transactionKey','transactionAmount','transactionType']]
Multiwipe_Du

customerId  ... transactionType
1        100088067  ...        PURCHASE
3        100737756  ...        PURCHASE
5        100737756  ...        PURCHASE
7        100737756  ...        PURCHASE
8        100737756  ...        PURCHASE
...            ...  ...             ...
105329   999257059  ...        PURCHASE
105331   999257059  ...        PURCHASE
105333   999257059  ...        PURCHASE
105335   999257059  ...        PURCHASE
105337   999283629  ...        PURCHASE

[68698 rows x 6 columns]

**Total dollar amount for the multi-swipe transactions is $10268522.62.**

In [26]:
Multiwipe_Du.sum()['transactionAmount']

10268522.62

###Delete the multi-wipe transactions

In [29]:
credit_card_no_duplicate=credit_card_no_reversed[~(credit_card_no_reversed.transactionKey.isin(Multiwipe_Du.transactionKey))]
credit_card_no_duplicate

accountNumber  customerId  ...  isFraud  transactionKey
0           737265056   737265056  ...    False              T0
1           737265056   737265056  ...    False              T1
2           737265056   737265056  ...    False              T2
3           737265056   737265056  ...    False              T3
4           830329091   830329091  ...    False              T4
...               ...         ...  ...      ...             ...
786358      732852505   732852505  ...    False         T786358
786359      732852505   732852505  ...    False         T786359
786360      732852505   732852505  ...    False         T786360
786361      732852505   732852505  ...    False         T786361
786362      732852505   732852505  ...    False         T786362

[682123 rows x 30 columns]

In [30]:
786363-682123

104240


*   I fond the total number of duplicate transactions was 104,240.
*   There were lots of monthly subscriptions transactions.



# **Data Processing**

In [31]:
# Drop columns with NaN values
process_data=credit_card_no_duplicate.drop(columns=['transactionDateTime','echoBuffer','merchantCity','merchantState','merchantZip','posOnPremises','recurringAuthInd'])
process_data

accountNumber  customerId  ...  isFraud  transactionKey
0           737265056   737265056  ...    False              T0
1           737265056   737265056  ...    False              T1
2           737265056   737265056  ...    False              T2
3           737265056   737265056  ...    False              T3
4           830329091   830329091  ...    False              T4
...               ...         ...  ...      ...             ...
786358      732852505   732852505  ...    False         T786358
786359      732852505   732852505  ...    False         T786359
786360      732852505   732852505  ...    False         T786360
786361      732852505   732852505  ...    False         T786361
786362      732852505   732852505  ...    False         T786362

[682123 rows x 23 columns]

## **Sampling**

Since this dataset is highly imbalanced, I retain the fraud data and sample from true transactions.
 

In [32]:
process_data['isFraud'].value_counts()# Imbalanced Data

False    671906
True      10217
Name: isFraud, dtype: int64

In [33]:
processed=process_data.copy()

In [34]:
# Find customers who had fraud transations
fraud_data=processed[processed.isFraud==True]
fraud_data.customerId.nunique()
processed=processed.reset_index(drop=True)
customer_data=fraud_data.customerId.unique() 

In [53]:
# Find true transactions with same customers
customer_fraud=processed[processed.customerId.isin(customer_data)]
true_tr=customer_fraud[customer_fraud.isFraud==False] 
df_false_1=true_tr.sample(frac=0.02,random_state=42)
df_false_1=df_false_1.reset_index(drop=True)
df_false_1

accountNumber  customerId  ...  isFraud  transactionKey
0          358749735   358749735  ...    False         T696646
1          699239145   699239145  ...    False         T514646
2          854042767   854042767  ...    False          T80703
3          770035360   770035360  ...    False         T687053
4          894938833   894938833  ...    False         T215349
...              ...         ...  ...      ...             ...
10850      259732658   259732658  ...    False         T564827
10851      119937133   119937133  ...    False         T563928
10852      118367557   118367557  ...    False         T450866
10853      338839426   338839426  ...    False         T109905
10854      681230823   681230823  ...    False          T77958

[10855 rows x 23 columns]

In [54]:
df_false_1.customerId.nunique()

1676

In [56]:
# Sample from other cusmtoers who did not have fraudulant transactions
true_data=processed[processed.isFraud==False]
df_new=true_data[~true_data.transactionKey.isin(true_tr.transactionKey)] # take transactions which 
df_new
df_false_2=df_new.sample(frac=0.08,random_state=42)
df_false_2=df_false_2.reset_index(drop=True)
df_false_2

accountNumber  customerId  ...  isFraud  transactionKey
0          167830614   167830614  ...    False         T594344
1          315850074   315850074  ...    False         T286603
2          144105109   144105109  ...    False         T545800
3          628652984   628652984  ...    False         T780361
4          887139069   887139069  ...    False         T372407
...              ...         ...  ...      ...             ...
10329      142462254   142462254  ...    False         T160744
10330      673695283   673695283  ...    False         T568623
10331      101394465   101394465  ...    False         T724633
10332      858587070   858587070  ...    False         T434332
10333      900317177   900317177  ...    False          T64761

[10334 rows x 23 columns]

In [57]:
credit_card=pd.concat([df_false_1,df_false_2,fraud_data],axis=0) # merge dataframes vertically
credit_card.isFraud.value_counts()

False    21189
True     10217
Name: isFraud, dtype: int64

## **Feature Engineering**

In [58]:
df_tr=credit_card.copy()
df_tr

accountNumber  customerId  ...  isFraud  transactionKey
0           358749735   358749735  ...    False         T696646
1           699239145   699239145  ...    False         T514646
2           854042767   854042767  ...    False          T80703
3           770035360   770035360  ...    False         T687053
4           894938833   894938833  ...    False         T215349
...               ...         ...  ...      ...             ...
785680      207667444   207667444  ...     True         T785680
785713      207667444   207667444  ...     True         T785713
785888      428856030   428856030  ...     True         T785888
786054      657364505   657364505  ...     True         T786054
786112      899818521   899818521  ...     True         T786112

[31406 rows x 23 columns]

In [59]:
# Match CVV 
df_tr['CVVInMatch']=np.where( credit_card['cardCVV'] == credit_card['enteredCVV'] , 'True', 'False')


In [60]:
# Drop some variables which are currenly unnecessary in my case
df_tr.drop(columns=['accountNumber','currentExpDate','accountOpenDate','dateOfLastAddressChange','cardLast4Digits'],inplace=True)
df_tr

customerId  creditLimit  ...  transactionKey  CVVInMatch
0        358749735      15000.0  ...         T696646        True
1        699239145      15000.0  ...         T514646        True
2        854042767       1000.0  ...          T80703        True
3        770035360       2500.0  ...         T687053        True
4        894938833      15000.0  ...         T215349        True
...            ...          ...  ...             ...         ...
785680   207667444       7500.0  ...         T785680        True
785713   207667444       7500.0  ...         T785713        True
785888   428856030        250.0  ...         T785888        True
786054   657364505      20000.0  ...         T786054        True
786112   899818521       2500.0  ...         T786112        True

[31406 rows x 19 columns]

In [61]:
#merchant variables
#merchantCategoryCode
category_group=pd.DataFrame(df_tr.groupby(['merchantCategoryCode','isFraud']).count()['transactionKey']).reset_index()
cf=category_group[category_group.isFraud==False][['merchantCategoryCode','transactionKey']]
cf.columns=['merchantCategoryCode','True_Count']
ct=pd.DataFrame(category_group.groupby(['merchantCategoryCode']).sum()['transactionKey']).reset_index()
ct.columns=['merchantCategoryCode','Total']
cf=cf.merge(ct,on='merchantCategoryCode',how='inner')
cf['Percent']=cf.True_Count/cf.Total
cf=cf.sort_values(by=['Percent'],ascending=True)
cf

merchantCategoryCode  True_Count  Total   Percent
0                airline         395    825  0.478788
14         online_retail        4448   8659  0.513685
13          online_gifts         869   1648  0.527306
17             rideshare        1143   2140  0.534112
1                   auto         475    728  0.652473
5                   food        1950   2910  0.670103
8              furniture         201    293  0.686007
3          entertainment        2374   3254  0.729564
18         subscriptions         657    860  0.763953
4               fastfood        3404   4420  0.770136
11                hotels         910   1143  0.796150
10                health         569    655  0.868702
16         personal care         616    693  0.888889
6          food_delivery         306    306  1.000000
12            mobileapps         773    773  1.000000
2            cable/phone          67     67  1.000000
15  online_subscriptions         601    601  1.000000
7                   fuel        1298   1298  1.000000
9                    gym         133    133  1.000000

In [62]:
def categ_merchant(x):
    if x in ['gym','fuel','hotels','online_subscriptions','cable/phone','mobileapps','food_delivery']:
        return 0
    elif x in ['personal care','health','hotels','fastfood','subscriptions','entertainment']:
        return 1
    elif x in ['furniture','food','auto','online_gifts','online_retails','rideshare','arline']:
        return 2

In [63]:
dummies_cm=pd.DataFrame(df_tr.merchantCategoryCode.apply(lambda x: categ_merchant(x)))
dummies_cm.columns=['Categ_merchant']
dummies_cm.Categ_merchant.value_counts()

1.0    9882
2.0    7719
0.0    4321
Name: Categ_merchant, dtype: int64

In [64]:
#merchantCountryCode
country_group=pd.DataFrame(df_tr.groupby(['merchantCountryCode','isFraud']).count()['transactionKey']).reset_index()
c_f=country_group[country_group.isFraud==False][['merchantCountryCode','transactionKey']]
c_f.columns=['merchantCountryCode','True_Count']
c_t=pd.DataFrame(country_group.groupby(['merchantCountryCode']).sum()['transactionKey']).reset_index()
c_t.columns=['merchantCountryCode','Total']
c_f=c_f.merge(c_t,on='merchantCountryCode',how='inner')
c_f['Percent']=c_f.True_Count/c_f.Total
c_f=c_f.sort_values(by=['Percent'],ascending=True)
c_f

merchantCountryCode  True_Count  Total   Percent
1                 MEX          71    120  0.591667
0                 CAN          77    121  0.636364
2                  PR          40     62  0.645161
3                  US       20984  31015  0.676576

In [65]:
def categ_country(x):
    if x in ['US']:
        return 0
    elif x in ['PR','CAN']:
        return 1
    elif x in ['MEX']:
        return 2

In [66]:
dummies_cc=pd.DataFrame(df_tr.merchantCountryCode.apply(lambda x: categ_country(x)))
dummies_cc.columns=['Categ_country']

In [67]:
df_data=pd.concat([dummies_cc,dummies_cm,df_tr],axis=1)
col =df_data.pop("CVVInMatch")
df_data.insert(5, 'CVVinMatch', col)
df_data.replace({False: 0, True: 1}, inplace=True)
df_data

Categ_country  Categ_merchant  ...  isFraud  transactionKey
0                 0.0             1.0  ...        0         T696646
1                 0.0             2.0  ...        0         T514646
2                 0.0             1.0  ...        0          T80703
3                 0.0             NaN  ...        0         T687053
4                 0.0             1.0  ...        0         T215349
...               ...             ...  ...      ...             ...
785680            0.0             NaN  ...        1         T785680
785713            0.0             1.0  ...        1         T785713
785888            0.0             NaN  ...        1         T785888
786054            0.0             2.0  ...        1         T786054
786112            0.0             NaN  ...        1         T786112

[31406 rows x 21 columns]

In [68]:
df_data.drop(columns=['customerId','merchantName','merchantCountryCode','acqCountry','merchantCategoryCode','transactionType','posEntryMode','posConditionCode'],inplace=True)
                                             

In [69]:
df_data.drop(columns=['currentBalance','transactionKey'],inplace=True)
df_data

Categ_country  Categ_merchant  ...  expirationDateKeyInMatch  isFraud
0                 0.0             1.0  ...                         0        0
1                 0.0             2.0  ...                         0        0
2                 0.0             1.0  ...                         0        0
3                 0.0             NaN  ...                         0        0
4                 0.0             1.0  ...                         0        0
...               ...             ...  ...                       ...      ...
785680            0.0             NaN  ...                         0        1
785713            0.0             1.0  ...                         0        1
785888            0.0             NaN  ...                         0        1
786054            0.0             2.0  ...                         0        1
786112            0.0             NaN  ...                         0        1

[31406 rows x 11 columns]

In [170]:
df_data.to_csv(abspath_curr+'df_data.csv')